File: concept/answer/ConceptMapGroup.java
package grakn.client.concept.answer;

import grakn.client.concept.Concept;
import grakn.client.concept.impl.ConceptImpl;
import grakn.protocol.AnswerProto;

import java.util.List;
import java.util.stream.Collectors;

public class ConceptMapGroup {
    private final Concept owner;
    private final List<ConceptMap> answers;

    public ConceptMapGroup(Concept owner, List<ConceptMap> answers) {
        this.owner = owner;
        this.answers = answers;
    }

    public static ConceptMapGroup of(AnswerProto.ConceptMapGroup e) {
        Concept owner = ConceptImpl.of(e.getOwner());
        List<ConceptMap> conceptMaps = e.getConceptMapsList().stream().map(ConceptMap::of).collect(Collectors.toList());
        return new ConceptMapGroup(owner, conceptMaps);
    }

    public Concept owner() {
        return this.owner;
    }

    public List<ConceptMap> answers() {
        return this.answers;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ConceptMapGroup a2 = (ConceptMapGroup) obj;
        return this.owner.equals(a2.owner) &&
                this.answers.equals(a2.answers);
    }

    @Override
    public int hashCode() {
        int hash = owner.hashCode();
        hash = 31 * hash + answers.hashCode();

        return hash;
    }
}

File: concept/answer/ConceptMapGroup.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import AnswerProto from "grakn-protocol/protobuf/answer_pb";
import {
    Concept,
    ConceptMap,
    ThingImpl,
    TypeImpl
} from "../../dependencies_internal";

export class ConceptMapGroup {
    private readonly _owner: Concept;
    private readonly _conceptMaps: ConceptMap[];

    constructor(owner: Concept, conceptMaps: ConceptMap[]) {
        this._owner = owner;
        this._conceptMaps = conceptMaps;
    }

    public static of(res: AnswerProto.ConceptMapGroup): ConceptMapGroup {
        let concept: Concept;
        if (res.getOwner().hasThing()) concept = ThingImpl.of(res.getOwner().getThing());
        else concept = TypeImpl.of(res.getOwner().getType());
        return new ConceptMapGroup(concept, res.getConceptMapsList().map((ans) => ConceptMap.of(ans)) as ConceptMap[])
    }

    owner(): Concept {
        return this._owner;
    }

    conceptMaps(): ConceptMap[] {
        return this._conceptMaps;
    }
}
